File: partition_to_k_equal_sum_subsets/PartitionToKEqualSumSubsets.java
class PartitionToKEqualSumSubsets {
    public boolean canPartitionKSubsets(int[] nums, int k) {
        if (nums == null || nums.length == 0 || nums.length < k) {
            return false;
        }

        long targetSum = 0;
        for (int i = 0; i < nums.length; i++) {
            targetSum += nums[i];
        }
        if (targetSum % k != 0) {
            return false;
        }
        
        return dfs(nums, new boolean[nums.length], 0, 0, k, 0, (int)(targetSum / (long)k));
    }

    private boolean dfs(int[] nums, boolean[] visited, int idx, int group, int targetGroup, int currSum, int targetSum) {
        if (group == targetGroup) {
            return true;
        }
        if (currSum == targetSum) {
            return dfs(nums, visited, 0, group + 1, targetGroup, 0, targetSum);
        }
        if (currSum > targetSum) {
            return false;
        }

        for (int i = idx; i < nums.length; i++) {
            if (visited[i]) {
                continue;
            }            
            visited[i] = true;
            if (dfs(nums, visited, i + 1, group, targetGroup, currSum + nums[i], targetSum)) {
                return true;
            }
            visited[i] = false;
        }

        return false;
    }
}

File: partition_to_k_equal_sum_subsets/Partition_To_K_Equal_Sum_Subsets.cpp
class Partition_To_K_Equal_Sum_Subsets {
public:
    bool canPartitionKSubsets(vector<int>& nums, int k) {
        if (nums.empty() || nums.size() < k) {
            return false;
        }

        long sum = 0;
        for (int num : nums) {
            sum += num;
        }
        if (sum % k != 0) {
            return false;
        }

        bool visited[nums.size()];
        for (int i = 0; i < nums.size(); i++) {
            visited[i] = false;
        }
        int target = (int)(sum / k);
        sort(nums.begin(), nums.end());
        return dfs(nums, visited, 0, 0, k, 0, target);
    }

private:
    bool dfs(vector<int>& nums, bool* visited, int idx, int group, int& target_group, int curr_sum, int& target_sum) {
        if (group == target_group) {
            return true;
        }
        if (curr_sum > target_sum) {
            return false;
        }
        if (curr_sum == target_sum) {
            return dfs(nums, visited, 0, group + 1, target_group, 0, target_sum);
        }

        int prev_num = INT_MIN;
        for (int i = idx; i < nums.size(); i++) {
            if (*(visited + i) || prev_num == nums[i]) {
                continue;
            }
            prev_num = nums[i];
            *(visited + i) = true;
            if (dfs(nums, visited, i + 1, group, target_group, curr_sum + nums[i], target_sum)) {
                return true;
            }
            *(visited + i) = false;
        }

        return false;
    }
};